### PR TITLE
Game Events - part 1

### PR DESCRIPTION
**Pull request by Zintixx**
Submitting this as a draft for feedback if my approach is appropriate.

### CODE REVIEW

File: Maple2.Model/Game/Event/GameEvent.cs [hunk 1]
```diff
@@ -1,16 +1,29 @@
...
+        writer.WriteUnicodeString(Name);
+    }
+}
 
-    protected GameEvent(string name) {
+public abstract class GameEventInfo : IByteSerializable {
+    public int Id;
+    public string Name;
+    protected GameEventInfo(string name) {
         Name = name;
```
Comment (kOchirasu) [inline Maple2.Model/Game/Event/GameEvent.cs:23]: Supposedly you can just used `Name = this.GetType()` and you will get the correct name of the implementing class and not require this to be passed in.

File: Maple2.Server.Game/PacketHandlers/EnterEventFieldHandler.cs [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+using Maple2.Server.Game.Session;
+
+namespace Maple2.Server.Game.PacketHandlers;
+
+public class EnterEventFieldHandler : PacketHandler<GameSession> {
+    public override RecvOp OpCode => RecvOp.EnterEventField;
+    
+    #region Autofac Autowired
+    // ReSharper disable MemberCanBePrivate.Global
+    public required GameStorage GameStorage { get; init; }
```
Comment (kOchirasu) [inline Maple2.Server.Game/PacketHandlers/EnterEventFieldHandler.cs:18]: `private get`

File: Maple2.Server.Game/PacketHandlers/EnterEventFieldHandler.cs [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+    // ReSharper restore All
+    #endregion
+    
+    public override void Handle(GameSession session, IByteReader packet) {
+        using GameStorage.Request db = GameStorage.Context();
+        GameEvent? gameEvent = db.FindEvent(nameof(EventFieldPopup));
+        if (gameEvent is not {EventInfo: EventFieldPopup fieldPopup}) {
```
Comment (kOchirasu) [inline Maple2.Server.Game/PacketHandlers/EnterEventFieldHandler.cs:25]: how about: `gameEvent.EventInfo is not EventFieldPopup fieldPopup`

File: Maple2.Server.Game/Packets/GameEventPacket.cs [hunk 4]
```diff
@@ -31,7 +32,7 @@ private enum Command : byte {
         pWriter.Write<Command>(Command.Unknown1);
         pWriter.WriteInt(gameEvents.Count);
         foreach (GameEvent gameEvent in gameEvents) {
-            pWriter.WriteClass<GameEvent>(gameEvent);
+            pWriter.WriteClass<GameEventInfo>(gameEvent.EventInfo);
```
Comment (kOchirasu) [inline Maple2.Server.Game/Packets/GameEventPacket.cs:35]: Here and below, you missed:
```cs
pWriter.WriteUnicodeString(gameEvent.Name);
```

File: Maple2.Database/Storage/Game/GameStorage.GameEvent.cs [hunk 5]
```diff
@@ -0,0 +1,24 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using Maple2.Model.Game.Event;
+
+namespace Maple2.Database.Storage;
+
+public partial class GameStorage {
+    public partial class Request {
+        public IEnumerable<GameEvent> GetEvents() {
```
Comment (kOchirasu) [inline Maple2.Database/Storage/Game/GameStorage.GameEvent.cs:10]: Probably fine to just return `IList<>` here.

```cs
return Context.GameEvent.Where(model => model.BeginTime <= DateTimeOffset.UtcNow && model.EndTime >= DateTimeOffset.UtcNow)
    .Select(model => {
        GameEvent result = model;
        result.EventInfo.Id = model.Id;
        result.EventInfo.Name = model.Name;
        return result;
    }).ToArray();
```

Comment (Zintixx) [reply Maple2.Database/Storage/Game/GameStorage.GameEvent.cs]: I wasn't able to do this. I needed to make the EventInfo's id and name defined. Not sure if there's a better way.

File: Maple2.Model/Game/Event/GameEvent.cs [hunk 6]
```diff
@@ -3,14 +3,24 @@
...
-        Name = name;
-    }
+public class GameEvent {
+    public int Id { get; init; }
+    public string Name { get; init; }
+    public long BeginTime { get; init; }
+    public long EndTime { get; init; }
+    public GameEventInfo EventInfo { get; init; }
 
     public virtual void WriteTo(IByteWriter writer) {
```
Comment (kOchirasu) [inline Maple2.Model/Game/Event/GameEvent.cs:13]: This is not `IByteSerializable` anymore so you don't need `WriteTo`

File: Maple2.Server.Game/PacketHandlers/UserChatHandler.cs [hunk 7]
```diff
@@ -135,6 +140,24 @@ public class UserChatHandler : PacketHandler<GameSession> {
     }
 
     private void HandleWorld(GameSession session, string message) {
+        var voucher = session.Item.Inventory.Filter(item => item.Metadata.Property.Tag == ItemTag.FreeWorldChatCoupon).FirstOrDefault();
+        if (voucher != null) {
+            session.Item.Inventory.Consume(voucher.Uid, 1);
```
Comment (kOchirasu) [inline Maple2.Server.Game/PacketHandlers/UserChatHandler.cs:145]: Can `Consume()` fail? We should handle the error.

File: Maple2.Database/Storage/Game/GameStorage.GameEvent.cs (hunk 5 above)
Comment (kOchirasu) [reply Maple2.Database/Storage/Game/GameStorage.GameEvent.cs]: I don't really understand why but it's okay. I'm not really against your current implementation.

**Pull request merged**